json.RawMessage does not support NULL

```
sql: Scan error on column index 5, name "xxx": unsupported Scan, storing driver.Value type <nil> into type *json.RawMessage
```

It appears that `json.RawMessage` does not implement the `Scanner` interface.

```
# command-line-arguments
./foo.go:24:21: n.RawMessage.Scan undefined (type json.RawMessage has no field or method Scan)
./foo.go:32:2: not enough arguments to return
./foo.go:32:21: n.RawMessage.Value undefined (type json.RawMessage has no field or method Value
```

--- a/internal/codegen/golang/imports.go
+++ b/internal/codegen/golang/imports.go
@@ -135,14 +135,17 @@ var stdlibTypes = map[string]string{
 var pgtypeTypes = map[string]struct{}{
 	"pgtype.CIDR":    {},
 	"pgtype.Inet":    {},
+	"pgtype.JSON":    {},
+	"pgtype.JSONB":   {},
 	"pgtype.Macaddr": {},
 	"pgtype.Numeric": {},
 }
 
 var pqtypeTypes = map[string]struct{}{
-	"pqtype.CIDR":    {},
-	"pqtype.Inet":    {},
-	"pqtype.Macaddr": {},
+	"pqtype.CIDR":           {},
+	"pqtype.Inet":           {},
+	"pqtype.Macaddr":        {},
+	"pqtype.NullRawMessage": {},
 }
 
 func buildImports(settings config.CombinedSettings, queries []Query, uses func(string) bool) (map[string]struct{}, map[ImportSpec]struct{}) {

--- a/internal/codegen/golang/postgresql_type.go
+++ b/internal/codegen/golang/postgresql_type.go
@@ -76,8 +76,33 @@ func postgresType(r *compiler.Result, col *compiler.Column, settings config.Comb
 		}
 		return "sql.NullBool"
 
-	case "json", "jsonb":
-		return "json.RawMessage"
+	case "json":
+		switch driver {
+		case SQLDriverPGXV4:
+			return "pgtype.JSON"
+		case SQLDriverLibPQ:
+			if notNull {
+				return "json.RawMessage"
+			} else {
+				return "pqtype.NullRawMessage"
+			}
+		default:
+			return "interface{}"
+		}
+
+	case "jsonb":
+		switch driver {
+		case SQLDriverPGXV4:
+			return "pgtype.JSONB"
+		case SQLDriverLibPQ:
+			if notNull {
+				return "json.RawMessage"
+			} else {
+				return "pqtype.NullRawMessage"
+			}
+		default:
+			return "interface{}"
+		}
 
 	case "bytea", "blob", "pg_catalog.bytea":
 		return "[]byte"

--- a/internal/endtoend/testdata/json/postgresql/pgx/go/models.go
+++ b/internal/endtoend/testdata/json/postgresql/pgx/go/models.go
@@ -3,12 +3,12 @@
 package querytest
 
 import (
-	"encoding/json"
+	"github.com/jackc/pgtype"
 )
 
 type Foo struct {
-	A json.RawMessage
-	B json.RawMessage
-	C json.RawMessage
-	D json.RawMessage
+	A pgtype.JSON
+	B pgtype.JSONB
+	C pgtype.JSON
+	D pgtype.JSONB
 }

--- a/internal/endtoend/testdata/json/postgresql/stdlib/go/models.go
+++ b/internal/endtoend/testdata/json/postgresql/stdlib/go/models.go
@@ -4,11 +4,13 @@ package querytest
 
 import (
 	"encoding/json"
+
+	"github.com/tabbed/pqtype"
 )
 
 type Foo struct {
 	A json.RawMessage
 	B json.RawMessage
-	C json.RawMessage
-	D json.RawMessage
+	C pqtype.NullRawMessage
+	D pqtype.NullRawMessage
 }

--- a/internal/endtoend/testdata/json_build/postgresql/pgx/go/query.sql.go
+++ b/internal/endtoend/testdata/json_build/postgresql/pgx/go/query.sql.go
@@ -5,7 +5,8 @@ package querytest
 
 import (
 	"context"
-	"encoding/json"
+
+	"github.com/jackc/pgtype"
 )
 
 const selectJSONBuildArray = `-- name: SelectJSONBuildArray :one
@@ -18,11 +19,11 @@ SELECT
 `
 
 type SelectJSONBuildArrayRow struct {
-	JsonBuildArray   json.RawMessage
-	JsonBuildArray_2 json.RawMessage
-	JsonBuildArray_3 json.RawMessage
-	JsonBuildArray_4 json.RawMessage
-	JsonBuildArray_5 json.RawMessage
+	JsonBuildArray   pgtype.JSON
+	JsonBuildArray_2 pgtype.JSON
+	JsonBuildArray_3 pgtype.JSON
+	JsonBuildArray_4 pgtype.JSON
+	JsonBuildArray_5 pgtype.JSON
 }
 
 func (q *Queries) SelectJSONBuildArray(ctx context.Context) (SelectJSONBuildArrayRow, error) {
@@ -48,11 +49,11 @@ SELECT
 `
 
 type SelectJSONBuildObjectRow struct {
-	JsonBuildObject   json.RawMessage
-	JsonBuildObject_2 json.RawMessage
-	JsonBuildObject_3 json.RawMessage
-	JsonBuildObject_4 json.RawMessage
-	JsonBuildObject_5 json.RawMessage
+	JsonBuildObject   pgtype.JSON
+	JsonBuildObject_2 pgtype.JSON
+	JsonBuildObject_3 pgtype.JSON
+	JsonBuildObject_4 pgtype.JSON
+	JsonBuildObject_5 pgtype.JSON
 }
 
 func (q *Queries) SelectJSONBuildObject(ctx context.Context) (SelectJSONBuildObjectRow, error) {

